drivers: usb_dc_rpi_pico: don't call usb_dc_ep_start_read() on transf…

Manu3l0us opened the pull request:
…er-managed endpoints during setup and properly reset next_pid.

The previous behaviour led to an issue where we already expected data1 on the first transfer instead of data0. The DesignWare USB DC actually implements the same behaviour. Also, the next_pid flag has to be reset on setting up the endpoint.

Fixes: #66283

jfischer-no commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -640,6 +640,7 @@ int usb_dc_ep_enable(const uint8_t ep)
 	/* clear buffer state (EP0 starts with PID=1 for setup phase) */
 
 	*ep_state->buf_ctl = (USB_EP_GET_IDX(ep) == 0 ? USB_BUF_CTRL_DATA1_PID : 0);
+	ep_state->next_pid = (USB_EP_GET_IDX(ep) == 0 ? 1 : 0);
```
This is a bit confusing (also the comment above, DATA0 PID follows SETUP PID), maybe it works because setup packet can always be received and driver below does not call udc_rpi_start_xfer() for control out endpoint and next PID is explicitly updated in udc_rpi_handle_setup(). Can you change these two lines, 642 and 643, to
```c

	*ep_state->buf_ctl = USB_BUF_CTRL_DATA0_PID;
	ep_state->next_pid = 0;
```
and test it?

Manu3l0us replied:
Okay, I changed it according to your suggestion and it works on my side.

jfischer-no replied:
But do **you** think it is the right way and makes sense? Not that you changed it just because I suggested it.

Manu3l0us replied:
It makes sense to me. As you said, next_pid for control is forced in udc_rpi_handle_setup. I had another close look at the code and I'm confident that this is the right way.

Commit:
drivers: usb_dc_rpi_pico: handling of data toggle after endpoint setup

The previous behaviour led to an issue where we already expected data1
on the first transfer instead of data0. The DesignWare USB DC actually
implements the same behaviour. Also, the next_pid flag has to be reset
on setting up the endpoint.
Fixes #66283.

Signed-off-by: Manuel Aebischer <manuel.aebischer@netmodule.com>

Manu3l0us commented on drivers/usb/device/usb_dc_rpi_pico.c:
```diff
@@ -652,7 +653,8 @@ int usb_dc_ep_enable(const uint8_t ep)
 
-	if (USB_EP_DIR_IS_OUT(ep) && ep != USB_CONTROL_EP_OUT) {
+	if ((USB_EP_DIR_IS_OUT(ep) && ep != USB_CONTROL_EP_OUT) &&
+	    (ep_state->cb != usb_transfer_ep_callback)) {
```
Fixed

jfischer-no reviewed (approved):
Thank you

Manu3l0us commented:
Is there still anything missing from my side or what's the status on this PR?
